docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,332 @@
+`java.io.OutputStream` to use `GZIPOutputStream` in similar fashion as above.
+`GZIPOutputStream` takes in a `java.io.OutputStream` which `ByteArrayOutputStream` is.
+
+Could also use function to convert stream to byte array but chances are will run out
+of memory doing so. Body will probably be large if using streaming trait on it.
+
+If that code to convert the types is not possible for some reason then will have to 
+handwrite gzip implementation.
+ 
+**/
+```
+
+### Implementation
+This feature is implemented by registering an [interceptor](interceptors.md) that replaces the outgoing request body with a compressed equivalent.
+
+The interceptor holds most of the logic for this feature to work. The search logic for requested to supported compression algorithms is inside it. The actual compression is initiated here and the header logic as well.
+```kotlin
+private val VALID_COMPRESSION_RANGE = 0..10485760
+private val DEFAULT_COMPRESSION_THRESHOLD_BYTES = 10240
+
+/**
+ * HTTP interceptor that compresses operation request payloads when eligible
+ */
+public class RequestCompressionTraitInterceptor(
+    private val compressionThreshold: Int = DEFAULT_COMPRESSION_THRESHOLD_BYTES,
+    private val requestedCompressionAlgorithms: List<String>,
+    private val supportedCompressionAlgorithms: List<CompressionAlgorithm>,
+) : HttpInterceptor {
+    
+    // Verify min compression size setting is in range
+    init {
+        require(compressionThreshold in VALID_COMPRESSION_RANGE) { "compressionThresholdBytes ($compressionThresholdBytes) must be in the range $VALID_COMPRESSION_RANGE" }
+    }
+
+    override suspend fun modifyBeforeRetryLoop(
+        context: ProtocolRequestInterceptorContext<Any, HttpRequest>
+    ): HttpRequest {
+        
+        // Determine if going forward with compression
+        val payloadSize = context.protocolRequest.body.contentLength
+        val streamingPayload = payloadSize == null
+        if (streamingPayload || payloadSize!! >= compressionThreshold) {
+
+            // Check if requested algorithm(s) is supported
+            supportedCompressionAlgorithms.find { supported ->
+                requestedCompressionAlgorithms.find { supported.id == it } != null
+            }?.let { algorithm ->
+
+                // Attempt compression
+                val request = context.protocolRequest.toBuilder()
+                request.body = request.body
+                // TODO: Write compression
+                request.headers.append("Content-Encoding", algorithm.id)
+                return request.build()
+            }
+        }
+        return context.protocolRequest
+    }
+}
+```
+
+### Testing
+Testing is done using both unit tests and Smithy HTTP protocol compliance tests. The SDK unit tests are used to verify behaviour and ensure it doesn't drift over time. AWS HTTP protocol tests are used to test the behavior in a more integrated way.
+
+### Appendix
+
+#### Codegen
```

Comment:
This doesn't really feel like Appendix content to me... I think it's a key part of the design (when to register the interceptor)